include/aws/common/aws_profile.h

```diff
@@ -0,0 +1,178 @@
+ * read profile collections from, and what types of credentials have priority.
+ *
+ * The profile specification is informally defined as "what the aws cli does" and
+ * formally defined in internal aws documents.
+ */
+struct aws_profile_property {
+    struct aws_allocator *allocator;
+    struct aws_string *name;
+    struct aws_string *value;
+    struct aws_hash_table sub_properties;
+    bool is_empty_valued;
+};
+
+struct aws_profile {
+    struct aws_allocator *allocator;
+    struct aws_string *name;
+    struct aws_hash_table properties;
+    bool has_profile_prefix;
+};
+
+/**
+ * The profile specification has rule exceptions based on what file
+ * the profile collection comes from.
+ */
+enum aws_profile_source_type { AWS_PST_NONE, AWS_PST_CONFIG, AWS_PST_CREDENTIALS };
+
+struct aws_profile_collection {
+    struct aws_allocator *allocator;
+    enum aws_profile_source_type profile_source;
+    struct aws_hash_table profiles;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/*************************
+ * Profile collection APIs
+ *************************/
+
+/**
+ * Clean up everything associated with a profile collection
+ */
+AWS_COMMON_API
+void aws_profile_collection_destroy(struct aws_profile_collection *profile_collection);
+
+/**
+ * Create a new profile collection by parsing a file with the specified path
+ */
+AWS_COMMON_API
+struct aws_profile_collection *aws_profile_collection_new_from_file(
+    struct aws_allocator *allocator,
+    const struct aws_string *file_path,
+    enum aws_profile_source_type source);
+
+/**
+ * Create a new profile collection by merging a config-file-based profile
+ * collection and a credentials-file-based profile collection
+ */
+AWS_COMMON_API
+struct aws_profile_collection *aws_profile_collection_new_from_merge(
+    struct aws_allocator *allocator,
+    const struct aws_profile_collection *config_profiles,
+    const struct aws_profile_collection *credentials_profiles);
+
+/**
+ * Create a new profile collection by parsing text in a buffer.  Primarily
+ * for testing.
+ */
+AWS_COMMON_API
+struct aws_profile_collection *aws_profile_collection_new_from_buffer(
+    struct aws_allocator *allocator,
+    const struct aws_byte_buf *buffer,
+    enum aws_profile_source_type source);
+
+/**
+ * Retrieves a profile with the specified name, if it exists, from the profile collection
+ */
+AWS_COMMON_API
+struct aws_profile *aws_profile_collection_get_profile(
+    const struct aws_profile_collection *profile_collection,
+    const struct aws_string *profile_name);
+
+/**
+ * Returns how many profiles a collection holds
+ */
+AWS_COMMON_API
+size_t aws_profile_collection_get_profile_count(const struct aws_profile_collection *profile_collection);
+
+/**************
+ * profile APIs
+ **************/
+
+/**
+ * Retrieves a property with the specified name, if it exists, from a profile
+ */
+AWS_COMMON_API
+struct aws_profile_property *aws_profile_get_property(
```

Comment:
document whether a copy, or a reference is being returned.

if it's a reference, we might want to add acquire/release semantics to these structs